p2panda-rs/src/permissions/key_group/key_group.rs

```diff
@@ -0,0 +1,297 @@
+    /// contain additional unrelated key groups.
+    pub fn new(
+        key_group: &KeyGroupView,
+        members: &[Membership],
+        member_key_groups: &[KeyGroup],
+    ) -> Result<KeyGroup, KeyGroupError> {
+        debug!("Building {}", key_group.name());
+
+        // Collect all (author, membership) pairs from `members` parameter, including duplicate
+        // author values.
+        let mut member_pool = vec![];
+        for membership in members {
+            match membership.member() {
+                // Simple case: for single key memberships just add that key to the pool.
+                Owner::Author(value) => {
+                    debug!("Adding author {:?}", value.as_str());
+                    member_pool.push((value, membership));
+                }
+
+                // When a key group is a member, recursively add that key group's members to the
+                // pool, assigned to a shared `membership`
+                Owner::KeyGroup(sub_key_group_id) => {
+                    let sub_key_group = member_key_groups
+                        .iter()
+                        .find(|key_group| key_group.id() == sub_key_group_id)
+                        .ok_or_else(|| {
+                            KeyGroupError::MissingMemberKeyGroup(format!("{:?}", sub_key_group_id))
+                        })?;
+
+                    debug!("Adding members of key group {}", sub_key_group.name());
+                    for author in sub_key_group.members().keys() {
+                        // Only add if a member is accepted within the sub key group.
+                        if sub_key_group.is_member(author) {
+                            debug!("Adding {}", author.as_str());
+                            member_pool.push((author, membership));
+                        } else {
+                            debug!("Skipping {}", author.as_str());
+                        }
+                    }
+                }
+            };
+        }
+
+        // Deduplicate so we have one membership per public key. A membership with more rights
+        // takes precedence here. At the moment that is just memberships that are accepted, vs not
+        // accepted.
+        let mut member_map: HashMap<Author, Vec<Membership>> = HashMap::new();
+        for (author, membership) in member_pool {
+            if let Some(previous_memberships) = member_map.get_mut(author) {
+                previous_memberships.push(membership.clone());
+            } else {
+                member_map.insert(author.clone(), vec![membership.clone()]);
+            }
+        }
+
+        let key_group = KeyGroup {
+            id: key_group.id().clone(),
+            view_id: key_group.view_id().clone(),
+            name: key_group.name().to_string(),
+            members: member_map,
+        };
+
+        Ok(key_group)
+    }
+
+    /// Create a new key group from predecessor documents.
```

Comment:
Not sure I understand what "predecessor" means in this comment?

Reply:
I'll rephrase that!